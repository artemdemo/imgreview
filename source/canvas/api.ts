/**
 * This is public API to manipulate with canvas components.
 * Canvas should be isolated from the interface components.
 */

import emitter from './events/eventsEmitter';
import * as keys from './events/eventsKeys';
import { TCreateArrowOptions } from './events/eventsTypes';

export type TImageData = {
    image: any;
    name: string;
};

export type TCanvasSize = {
    width: number;
    height: number;
};

// Emitting events
//

export const createArrow = (options?: TCreateArrowOptions) => {
    emitter.emit(keys.CREATE_ARROW, options);
};

export const setImage = (data: TImageData) => {
    emitter.emit(keys.SET_IMAGE, data);
};

export const exportCanvasToImage = (name: string) => {
    emitter.emit(keys.EXPORT_CANVAS_TO_IMAGE, name);
};

export const updateCanvasSize = (data: TCanvasSize) => {
    emitter.emit(keys.UPDATE_CANVAS_SIZE, data);
};

export const blurShapes = () => {
    emitter.emit(keys.BLUR_SHAPES);
};

// Subscribing to events
//

export const onImageUpdate = (cb: () => void) => {
    emitter.on(keys.ON_IMAGE_UPDATE, cb);
};
