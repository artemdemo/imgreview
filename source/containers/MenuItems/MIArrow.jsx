import React from 'react';
import { connect } from 'react-redux';
import Icon from '../../components/Icon/Icon';
import MainMenuItem from '../../components/MainMenu/MainMenuItem';
import Arrow from '../../canvas/Arrow/Arrow';
import { addArrow, blurShapes } from '../../model/shapes/shapesActions';

class MIArrow extends React.PureComponent {
    onClick = () => {
        const { shapes, canvas, addArrow, blurShapes } = this.props;
        const arrow = new Arrow({
            stroke: shapes.stroke,
            strokeWidth: shapes.strokeWidth,
        });
        arrow.addToStage(canvas.stage);
        arrow.on('click', arrowInstance => blurShapes(arrowInstance));
        addArrow(arrow);
    };

    render() {
        const { canvas } = this.props;
        return (
            <MainMenuItem
                onClick={this.onClick}
                disabled={canvas.image == null}
            >
                <Icon name='mouse-pointer' />
            </MainMenuItem>
        );
    }
}

export default connect(
    state => ({
        canvas: state.canvas,
        shapes: state.shapes,
    }),
    {
        addArrow,
        blurShapes,
    }
)(MIArrow);
