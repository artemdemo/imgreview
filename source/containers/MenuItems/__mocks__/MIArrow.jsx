import React from 'react';

const MIArrow = props => (
    <div data-mock='MIArrow'>
        {props.children}
    </div>
);

export default MIArrow;
