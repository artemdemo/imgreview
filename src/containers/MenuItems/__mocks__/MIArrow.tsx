import React from 'react';

const MIArrow = props => (
    <div data-mock='MIArrow'>
        {JSON.stringify(props, null, 2)}
    </div>
);

export default MIArrow;
