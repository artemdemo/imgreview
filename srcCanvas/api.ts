/**
 * This is public API to manipulate with canvas components.
 * Canvas should be isolated from the main application.
 */

import { createNanoEvents } from "nanoevents";
import { createEvent } from "./events/eventCreator";
import { TCreateArrowOptions, TCreateTextOptions } from "./events/eventsTypes";
export { default as shapeTypes } from "./Shape/shapeTypes";

const emitter = createNanoEvents();

export type TImageData = {
    image: any;
    name: string;
};

export type TCanvasSize = {
    width: number;
    height: number;
};

// Emitting events
//

export const createArrow: (options?: TCreateArrowOptions) => void = createEvent(emitter, 'CREATE_ARROW');

export const createText: (options?: TCreateTextOptions) => void = createEvent(emitter, 'CREATE_TEXT');

export const setImage: (data: TImageData) => void = createEvent(emitter, 'SET_IMAGE');

export const setStrokeColorToActiveShape: (hex: string) => void = createEvent(emitter, 'SET_STROKE_COLOR_TO_ACTIVE_SHAPE');

export const setStrokeWidthToActiveShape: (width: number) => void = createEvent(emitter, 'SET_STROKE_WIDTH_TO_ACTIVE_SHAPE');

export const setFontSizeToActiveShape: (width: number) => void = createEvent(emitter, 'SET_FONT_SIZE_TO_ACTIVE_SHAPE');

export const exportCanvasToImage : (name: string) => void = createEvent(emitter, 'EXPORT_CANVAS_TO_IMAGE');

export const updateCanvasSize: (data: TCanvasSize) => void = createEvent(emitter, 'UPDATE_CANVAS_SIZE');

export const blurShapes: () => void = createEvent(emitter, 'BLUR_SHAPES');

// Create blank canvas, it's an easy way to test shapes,
// without opening an image
export const initBlankCanvas: (props: { width: number, height: number }) => void = createEvent(emitter, 'INIT_BLANK_CANVAS');

// Subscribing to events
//

export const imageUpdated: (size: number) => void = createEvent(emitter, 'IMAGE_UPDATED');

export const shapeClicked: (shape: any) => void = createEvent(emitter, 'SHAPE_CLICKED');

export const shapesBlurred: (shape: any) => void = createEvent(emitter, 'SHAPES_BLURRED');

export const shapeAdded: (shape: any) => void = createEvent(emitter, 'SHAPE_ADDED');
