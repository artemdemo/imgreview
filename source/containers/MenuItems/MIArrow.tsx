import React from 'react';
import { connect } from 'react-redux';
import Icon from '../../components/Icon/Icon';
import TopMenuItem from '../../components/TopMenu/TopMenuItem';
import { createArrow } from '../../canvas/api';
import { TReduxState } from '../../reducers';
import { TStateShapes } from '../../model/shapes/shapesReducer';

type Props = {
    disabled: boolean;
    shapes: TStateShapes,
};

class MIArrow extends React.PureComponent<Props> {
    static readonly defaultProps = {
        disabled: false,
    };

    onClick = () => {
        const { shapes } = this.props;
        createArrow({
            strokeColor: shapes.strokeColor,
            strokeWidth: shapes.strokeWidth,
        });
    };

    render() {
        const { disabled } = this.props;
        return (
            <TopMenuItem
                onClick={this.onClick}
                disabled={disabled}
            >
                <Icon
                    name='mouse-pointer'
                    title='Arrow'
                />
            </TopMenuItem>
        );
    }
}

export default connect(
    (state: TReduxState) => ({
        shapes: state.shapes,
    })
)(MIArrow);
