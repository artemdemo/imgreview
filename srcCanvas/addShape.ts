/* eslint-disable import/prefer-default-export */

import _ from 'lodash';
import canvasStore from './store';
import {addShape, blurShapes, setCursor} from './model/shapes/shapesActions';
import {ECursorTypes} from './model/shapes/shapesModelTypes';
import CanvasImage from './canvasShapes/Image/CanvasImage';
import Arrow from './canvasShapes/Arrow/Arrow';
import Text from './canvasShapes/Text/Text';
import * as canvasApi from './api';
import {TCanvasState} from './reducers';
import {
  TCreateArrowOptions,
  TCreateEllipseOptions,
  TCreateRectOptions,
  TCreateTextOptions,
} from './events/eventsTypes';
import Rect from './canvasShapes/RectLike/Rect';
import Shape from './canvasShapes/Shape/Shape';
import SelectRect from './canvasShapes/RectLike/SelectRect';
import EShapeTypes from './canvasShapes/Shape/shapeTypes';
import Ellipse from './canvasShapes/RectLike/Ellipse';
import RectRough from './canvasShapes/RectLike/RectRough';
import EllipseRough from './canvasShapes/RectLike/EllipseRough';

/**
 * Add standard events to the shape.
 * @param shape
 */
const attachGeneralEvents = (shape: Shape) => {
  shape.on('click', (shapeInstance) => {
    canvasStore.dispatch(blurShapes(shapeInstance));
  });
  shape.on('dragstart', (shapeInstance) => {
    canvasStore.dispatch(blurShapes(shapeInstance));
  });
  shape.on('mouseover', () => {
    const { shapes, stage } = <TCanvasState>canvasStore.getState();
    if (stage.isDraggable) {
      shape.draggable(false);
    } else {
      // While adding shape user shouldn't be able to interact with existing shapes.
      shape.draggable(!shapes.addingShapeRef);
    }
    const cursor = shapes.addingShapeRef
      ? ECursorTypes.AUTO
      : ECursorTypes.MOVE;
    // Cursor should be changed only if stage is not dragged.
    // In this case, cursor is already set.
    if (!stage.isDraggable) {
      canvasStore.dispatch(setCursor(cursor));
    }
  });
  shape.on('mouseout', () => {
    const { stage } = <TCanvasState>canvasStore.getState();
    // Cursor should be changed only if stage is not dragged.
    // In this case, cursor is already set.
    if (!stage.isDraggable) {
      canvasStore.dispatch(setCursor(ECursorTypes.AUTO));
    }
  });
  const unsubShapeAdded = canvasApi.shapeAdded.on(() => {
    shape.draggable(true);
  });
  shape.on('_beforedestroy', () => {
    unsubShapeAdded();
  });
};

export const _createArrow = (
  arrow?: Arrow,
  options?: TCreateArrowOptions,
): Arrow => {
  const _arrow =
    arrow ||
    new Arrow({
      stroke: _.get(options, 'strokeColor', 'green'),
      strokeWidth: _.get(options, 'strokeWidth'),
    });
  _arrow.onAnchor('mouseover', () =>
    canvasStore.dispatch(setCursor(ECursorTypes.POINTER)),
  );
  _arrow.onAnchor('mouseout', () =>
    canvasStore.dispatch(setCursor(ECursorTypes.AUTO)),
  );
  attachGeneralEvents(_arrow);
  return _arrow;
};

export const _connectArrow = (arrow: Arrow) => {
  canvasStore.dispatch(addShape(arrow));
};

/**
 * Create and connect Arrow to the stage.
 * If arrow provided - it will use provided instance,
 * if not - will create new one.
 * @param arrow {Arrow} - I'm using it when coping Arrows.
 * @param options {object}
 */
export const createAndConnectArrow = (
  arrow?: Arrow,
  options?: TCreateArrowOptions,
) => {
  const _arrow = _createArrow(arrow, options);
  _connectArrow(_arrow);
};

export const _createText = (
  textNode?: Text,
  options?: TCreateTextOptions,
): Text => {
  const _textNode =
    textNode ||
    new Text({
      fill: _.get(options, 'fillColor', 'green'),
      fontSize: _.get(options, 'fontSize'),
    });
  _textNode.onDblClickGetStagePosition(() => {
    const { stage } = <TCanvasState>canvasStore.getState();
    const stageBox = stage.instance?.getBoundingClientRect();
    return {
      left: stageBox ? stageBox.left : 0,
      top: stageBox ? stageBox.top : 0,
    };
  });
  attachGeneralEvents(_textNode);
  return _textNode;
};

export const _connectText = (textNode: Text) => {
  canvasStore.dispatch(addShape(textNode));
};

export const createAndConnectText = (
  textNode?: Text,
  options?: TCreateTextOptions,
) => {
  const _textNode = _createText(textNode, options);
  _connectText(_textNode);
};

type TRectLike = Rect | RectRough | SelectRect | Ellipse;
type TCreateRectLikeOptions = TCreateRectOptions | TCreateEllipseOptions;

export const _createRectLike = (
  rectNode?: TRectLike,
  options?: TCreateRectLikeOptions,
  type?: EShapeTypes,
): TRectLike => {
  const props = {
    stroke: _.get(options, 'strokeColor', 'green'),
    fill: _.get(options, 'fill', 'transparent'),
    strokeWidth: _.get(options, 'strokeWidth', 2),
  };
  const _rectNode = (() => {
    if (rectNode) {
      return rectNode;
    }
    switch (type) {
      case EShapeTypes.RECT:
        return new Rect(props);
      case EShapeTypes.RECT_ROUGH:
        return new RectRough(props);
      case EShapeTypes.ELLIPSE:
        return new Ellipse(props);
      case EShapeTypes.ELLIPSE_ROUGH:
        return new EllipseRough(props);
      case EShapeTypes.SELECT_RECT:
        return new SelectRect();
    }
  })();
  if (!_rectNode) {
    throw new Error('rectNode is not defined');
  }
  attachGeneralEvents(_rectNode);
  return _rectNode;
};

export const _connectRectLike = (rectLikeNode: TRectLike) => {
  canvasStore.dispatch(addShape(rectLikeNode));
};

export const createAndConnectRectLike = (
  rectLikeNode?: TRectLike,
  options?: TCreateRectOptions,
  type?: EShapeTypes,
) => {
  const rect = _createRectLike(rectLikeNode, options, type);
  _connectRectLike(rect);
};

export const createAndConnectSelectRect = () => {
  const selectRectNode = _createRectLike();
  _createRectLike(selectRectNode);
};

export const connectShape = (shape: Shape) => {
  switch (shape.type) {
    case EShapeTypes.ARROW:
      _connectArrow(<Arrow>shape);
      break;
    case EShapeTypes.TEXT:
      _connectText(<Text>shape);
      break;
    case EShapeTypes.RECT:
    case EShapeTypes.RECT_ROUGH:
    case EShapeTypes.ELLIPSE:
    case EShapeTypes.ELLIPSE_ROUGH:
    case EShapeTypes.SELECT_RECT:
      _connectRectLike(<Rect | Ellipse>shape);
      break;
    default:
      console.error("Can't connect given shape");
      console.log(shape);
  }
};

/**
 * This is general function that will be used for connecting pasted shapes.
 * Usage will be in CanvasEl.tsx
 * @param shape {Shape}
 * @param options
 */
export const cloneAndConnectShape = (shape: Shape, options?: any) => {
  switch (shape.type) {
    case EShapeTypes.ARROW:
      // Here I'm copying again (first time was in `shapesReducer`),
      // this way user could paste shape multiple times without collisions
      createAndConnectArrow((<Arrow>shape).clone(), options);
      break;
    case EShapeTypes.TEXT:
      createAndConnectText((<Text>shape).clone(), options);
      break;
    case EShapeTypes.RECT:
    case EShapeTypes.RECT_ROUGH:
    case EShapeTypes.ELLIPSE:
    case EShapeTypes.ELLIPSE_ROUGH:
      createAndConnectRectLike((<Rect | Ellipse>shape).clone(), options);
      break;
    case EShapeTypes.IMAGE:
      connectImage((<CanvasImage>shape).clone());
      break;
    default:
      console.error("Can't clone and connect given shape");
      console.log(shape);
  }
};

const connectImage = (canvasImage: CanvasImage) => {
  attachGeneralEvents(canvasImage);
  canvasStore.dispatch(addShape(canvasImage));
};

/**
 * Add Image to stage
 * @param data {object}
 */
export const addImageToStage = (data: canvasApi.SetImageData) => {
  const { stage } = canvasStore.getState() as TCanvasState;
  if (!stage.instance) {
    throw new Error('Stage is not defined');
  }
  const absPos = stage.instance.absolutePosition()!;
  const canvasImage = new CanvasImage(
    data.image,
    data.pos
      ? {
          x: data.pos.x - absPos.x,
          y: data.pos.y - absPos.y,
        }
      : {
          x: stage.instance.width() / 2 - data.image.width / 2 - absPos.x,
          y: stage.instance.height() / 2 - data.image.height / 2 - absPos.y,
        },
  );
  connectImage(canvasImage);
};
