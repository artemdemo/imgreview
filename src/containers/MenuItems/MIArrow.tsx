import React from 'react';
import {connect} from 'react-redux';
import {FontAwesomeIcon} from '@fortawesome/react-fontawesome';
import {faLongArrowAltUp} from '@fortawesome/free-solid-svg-icons';
import TopMenuItem from '../../components/TopMenu/TopMenuItem';
import * as canvasApi from '../../../srcCanvas/api';
import {TReduxState} from '../../reducers';
import {TStateMenu} from '../../model/menu/menuReducer';
import {setShapeToAdd, TSetShapeToAdd} from '../../model/menu/menuActions';
import * as gaService from '../../services/ganalytics';
import { t } from '../../services/i18n';

type TProps = {
    disabled: boolean;
    menu: TStateMenu;
    setShapeToAdd: TSetShapeToAdd;
};

type TState = {
    active: boolean;
};

class MIArrow extends React.PureComponent<TProps, TState> {
    static readonly defaultProps = {
        disabled: false,
    };

    #unsubShapeAdded;

    state = {
        active: false,
    };

    componentDidMount() {
        // @ts-ignore
        this.#unsubShapeAdded = canvasApi.shapeAdded.on(this.handleShapeAdded);
    }

    componentWillUnmount() {
        this.#unsubShapeAdded();
    }

    handleShapeAdded = () => {
        this.setState({
            active: false,
        });
    };

    onClick = () => {
        const { menu, setShapeToAdd } = this.props;
        canvasApi.startAddingShape({
            type: canvasApi.EShapeTypes.ARROW,
            options: {
                strokeColor: menu.strokeColor,
                strokeWidth: menu.strokeWidth,
            },
        });

        setShapeToAdd(canvasApi.EShapeTypes.ARROW);

        gaService.sendEvent({
            eventCategory: gaService.EEventCategories.MenuClick,
            eventAction: gaService.EEventActions.AddArrow,
        });
    };

    render() {
        const { disabled, menu } = this.props;
        return (
            <TopMenuItem
                onClick={this.onClick}
                disabled={disabled}
                active={menu.selectedShapeToAdd === canvasApi.EShapeTypes.ARROW}
                title={t('menu.addArrow')}
            >
                <FontAwesomeIcon icon={faLongArrowAltUp} rotation={270} />
            </TopMenuItem>
        );
    }
}

export default connect(
    (state: TReduxState) => ({
        menu: state.menu,
    }), {
        setShapeToAdd,
    },
)(MIArrow);
