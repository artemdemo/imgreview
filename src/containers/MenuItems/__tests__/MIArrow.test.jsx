import React from "react";
import { mount } from "enzyme";
import renderer from "react-test-renderer";
import MIArrow from "../MIArrow.tsx";
import * as canvasApi from '../../../../srcCanvas/api';

jest.mock('react-redux');
jest.mock('../../../components/TopMenu/TopMenuItem');
jest.mock('../../../../srcCanvas/api');
jest.mock('../../../services/ganalytics');

describe('MIArrow', () => {
    const canvasApi = require('../../../../srcCanvas/api');
    const ganalyticsMock = require('../../../services/ganalytics');

    beforeAll(() => {
        jest.clearAllMocks();
    });

    it('should render', () => {
        const tree = renderer.create(
            <MIArrow />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should be disabled', () => {
        const tree = renderer.create(
            <MIArrow
                disabled
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should handle click', () => {
        const strokeColor = 'red';
        const strokeWidth = 5;
        const wrapper = mount(
            <MIArrow
                menu={{
                    strokeColor,
                    strokeWidth,
                }}
            />
        );
        wrapper.simulate('click');
        expect(canvasApi.startAddingShape).toBeCalledWith(
            canvasApi.EShapeTypes.ARROW,
            {
                strokeColor,
                strokeWidth,
            },
        );
        expect(ganalyticsMock.sendEvent).toBeCalledWith({
            eventAction: 'add Arrow',
            eventCategory: 'MenuClick',
        });
    });
});
