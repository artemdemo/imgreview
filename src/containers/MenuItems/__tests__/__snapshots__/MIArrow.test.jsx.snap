// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MIArrow should be disabled 1`] = `
<div
  data-mock="TopMenuItem"
  onClick={[Function]}
>
  <span
    data-mock="FontAwesomeIcon"
  >
    {
  "icon": "faLongArrowAltUp",
  "rotation": 270
}
  </span>
</div>
`;

exports[`MIArrow should render 1`] = `
<div
  data-mock="TopMenuItem"
  onClick={[Function]}
>
  <span
    data-mock="FontAwesomeIcon"
  >
    {
  "icon": "faLongArrowAltUp",
  "rotation": 270
}
  </span>
</div>
`;
